pallets/external-validators/src/lib.rs

```diff
@@ -0,0 +1,697 @@
+            let active_era = ActiveEra::<T>::mutate(|active_era| {
+                let new_index = active_era.as_ref().map(|info| info.index + 1).unwrap_or(0);
+                *active_era = Some(ActiveEraInfo {
+                    index: new_index,
+                    // Set new active era start in next `on_finalize`. To guarantee usage of `Time`
+                    start: None,
+                });
+                new_index
+            });
+            T::OnEraStart::on_era_start(active_era, start_session);
+        }
+
+        /// Compute payout for era.
+        pub(crate) fn end_era(active_era: ActiveEraInfo, _session_index: SessionIndex) {
+            // Note: active_era_start can be None if end era is called during genesis config.
+            if let Some(_active_era_start) = active_era.start {
+                // TODO: EraPaid event here
+            }
+            T::OnEraEnd::on_era_end(active_era.index);
+        }
+
+        /// Plan a new era.
+        ///
+        /// * Bump the current era storage (which holds the latest planned era).
+        /// * Store start session index for the new planned era.
+        /// * Clean old era information.
+        /// * Store staking information for the new planned era
+        ///
+        /// Returns the new validator set.
+        pub fn trigger_new_era(start_session_index: SessionIndex) -> Vec<T::ValidatorId> {
+            // Increment or set current era.
+            let new_planned_era = CurrentEra::<T>::mutate(|s| {
+                *s = Some(s.map(|s| s + 1).unwrap_or(0));
+                s.unwrap()
+            });
+            ErasStartSessionIndex::<T>::insert(&new_planned_era, &start_session_index);
+
+            // Clean old era information.
+            if let Some(old_era) = new_planned_era.checked_sub(T::HistoryDepth::get() + 1) {
+                Self::clear_era_information(old_era);
+            }
+
+            // Returns new validators
+            Self::validators()
+        }
+
+        /// Potentially plan a new era.
+        ///
+        /// Get election result from `T::ElectionProvider`.
+        /// In case election result has more than [`MinimumValidatorCount`] validator trigger a new era.
+        ///
+        /// In case a new era is planned, the new validator set is returned.
+        pub(crate) fn try_trigger_new_era(
+            start_session_index: SessionIndex,
+        ) -> Option<Vec<T::ValidatorId>> {
+            Some(Self::trigger_new_era(start_session_index))
+        }
+
+        /// Clear all era information for given era.
+        pub(crate) fn clear_era_information(era_index: EraIndex) {
+            ErasStartSessionIndex::<T>::remove(era_index);
+        }
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+        fn on_initialize(_now: BlockNumberFor<T>) -> Weight {
+            // just return the weight of the on_finalize.
+            T::DbWeight::get().reads(1)
```

Comment:
this should be probably 2 reads and one write:
- one read for ActiveEra
- one read for the Timestamp
- one weire for the Active era modification

Reply:
Ignoring this because I just copied it from polkadot

Reply:
ignore